File: apps/windows/TeslaSync.App/feature-views/AddWidgetButton/AddWidgetButton.Model.cs
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The render-time data model the <c>AddWidgetButton</c> surface binds to — the native analogue of the web
/// <c>AddWidgetButtonProps</c> (<c>{ onClick, isEditing }</c> in
/// web/src/features/dashboard/components/AddWidgetButton.tsx). The web component is a purely presentational,
/// dependency-free fragment: it fetches nothing, so there is no loading / empty / error / stale / offline
/// branch to reproduce here (those belong to data-backed surfaces). Only the <see cref="IsEditing"/> flag
/// drives what renders — when the dashboard is in edit mode the floating action button hides (web
/// <c>return null</c>), because the dashboard header already exposes its own "Add Widget" action. The click
/// callback (web <c>onClick</c>) is modelled as the view's <c>Click</c> event rather than a field, so this
/// model stays a pure, UI-free value. Unit-tested without a UI host.
/// </summary>
/// <param name="IsEditing">Whether the dashboard is in edit mode (web <c>isEditing</c>); hides the FAB when true.</param>
public sealed record AddWidgetButtonModel(bool IsEditing)
{
    /// <summary>The default model — the FAB shown (the dashboard is not in edit mode).</summary>
    public static AddWidgetButtonModel Visible { get; } = new(false);

    /// <summary>The edit-mode model — the FAB hidden (the header owns the add action).</summary>
    public static AddWidgetButtonModel Editing { get; } = new(true);
}

/// <summary>
/// The fully projected, render-ready view of an <see cref="AddWidgetButtonModel"/> — everything the web
/// component derives before returning JSX: whether the floating button is shown at all (web
/// <c>if (isEditing) return null</c>), the localized <see cref="Label"/> (web
/// <c>t('dashboard.addWidget', 'Add Widget')</c>) reused for the tooltip content and the accessible name, and
/// the composed Narrator <see cref="AutomationName"/>. Pure data so every value is asserted headlessly.
/// </summary>
/// <param name="IsVisible">True when the FAB renders; false in edit mode (web <c>return null</c>).</param>
/// <param name="Label">The resolved button label.</param>
/// <param name="AutomationName">The Narrator name for the button (web <c>aria-label</c>).</param>
/// <param name="TooltipHint">The tooltip text shown on hover/focus (web <c>Tooltip content</c>).</param>
public sealed record AddWidgetButtonDisplay(
    bool IsVisible,
    string Label,
    string AutomationName,
    string TooltipHint);

/// <summary>
/// Pure projection from an <see cref="AddWidgetButtonModel"/> to its <see cref="AddWidgetButtonDisplay"/> —
/// the native port of web/src/features/dashboard/components/AddWidgetButton.tsx. Reproduces the web
/// derivations exactly: visibility is <c>!isEditing</c> (the web early <c>return null</c>), and the single
/// label is the resolved <c>dashboard.addWidget</c> string, reused verbatim for the tooltip content (web
/// <c>Tooltip content</c>) and the button <c>aria-label</c>. No WinUI types — unit-tested without a UI host.
/// </summary>
public static class AddWidgetButtonProjection
{
    /// <summary>Project <paramref name="model"/> into a render-ready display, resolving copy via <paramref name="localizer"/>.</summary>
    /// <param name="model">The render-time data model (the web props minus the click callback).</param>
    /// <param name="localizer">The i18n facade resolving the single label key.</param>
    public static AddWidgetButtonDisplay Project(AddWidgetButtonModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        var label = AddWidgetButtonRegistration.Label(localizer);
        return new AddWidgetButtonDisplay(
            IsVisible: !model.IsEditing,
            Label: label,
            AutomationName: label,
            TooltipHint: label);
    }
}

/// <summary>
/// Canonical metadata for the <c>AddWidgetButton</c> feature surface — the native mirror of the web component
/// at web/src/features/dashboard/components/AddWidgetButton.tsx: the stable diagnostics slug, the single i18n
/// key (with the same English fallback the web <c>t(...)</c> call carries) and the Segoe Fluent Icons glyph
/// that stands in for the web Lucide <c>Plus</c> icon. UI-free so the metadata is asserted in tests.
/// </summary>
public static class AddWidgetButtonRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11).</summary>
    public const string Slug = "AddWidgetButton";

    /// <summary>The i18n key for the button label (web <c>t('dashboard.addWidget', ...)</c>).</summary>
    public const string LabelKey = "dashboard.addWidget";

    /// <summary>The English fallback the web <c>t(...)</c> call carries for <see cref="LabelKey"/>.</summary>
    public const string LabelFallback = "Add Widget";

    /// <summary>Segoe Fluent "Add" glyph — the native stand-in for the web Lucide <c>Plus</c> icon.</summary>
    public const string AddGlyph = "\uE710";

    /// <summary>The localized button label (tooltip content + accessible name).</summary>
    public static string Label(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer.GetString(LabelKey, LabelFallback);
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>AddWidgetButton</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event and the data-free catalogue-open activation with the surface slug —
/// never any dashboard or widget content — so a diagnostics line can never leak fleet state. Thread-safe.
/// </summary>
public sealed class AddWidgetButtonDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;
    private long _activations;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public AddWidgetButtonDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Number of times the FAB has been activated (the widget catalogue requested).</summary>
    public long Activations => Interlocked.Read(ref _activations);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=AddWidgetButton</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={AddWidgetButtonRegistration.Slug}");
    }

    /// <summary>Record that the FAB was activated, emitting <c>add-widget.activated slug=AddWidgetButton</c>.</summary>
    public void RecordActivated()
    {
        Interlocked.Increment(ref _activations);
        _sink?.Invoke($"add-widget.activated slug={AddWidgetButtonRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/AddWidgetButton/AddWidgetButtonModel.kt
// Pure, framework-free model + projection for the AddWidgetButton feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/dashboard/components/AddWidgetButton.tsx). No Compose, no Android, no HTTP: every
// declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// AddWidgetButton is the dashboard's floating "+" action: the web component takes an `onClick` callback and
// an `isEditing` flag as props from the Dashboard page, and renders a tooltip-wrapped circular primary FAB
// that opens the widget-catalogue dialog — except in edit mode, where it returns null because the dashboard
// header already exposes an "Add Widget" action. So the two branches the web source defines — not-editing
// (show the FAB) and editing (render nothing) — are the complete state set this surface renders. Exactly as
// the sibling WeekSelector / StatusHeader / SummaryStatsRow presentational ports document, the
// loading / empty / error / stale / offline states live on the owning page (which owns the dashboard query
// and the edit-mode client state), not here; the only data source the web component itself binds is
// `useTranslation`, mapped natively to the generated i18n catalog (P1/S10).
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/AddWidgetButton — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.addwidgetbutton

import io.teslasync.shared.core.diagnostics.Logger

/**
 * The mutually-exclusive surface the button renders, derived by [AddWidgetButtonProjection]. Each maps to a
 * render branch in the composable so the control is never ambiguous:
 *  - [Visible] — the dashboard is not in edit mode: the tooltip-wrapped "+" FAB (web content branch).
 *  - [Hidden]  — the dashboard is in edit mode: the surface renders nothing, because the dashboard header
 *    already exposes an "Add Widget" action (web `if (isEditing) return null`).
 */
enum class AddWidgetButtonSurface {
    Visible,
    Hidden,
}

/**
 * The render-ready view — the native analogue of the single decision the web component makes before
 * returning JSX (`isEditing ? null : <Fab/>`). Pure data (no Compose types) so the projection is fully
 * covered by the off-device unit gate.
 *
 * @property surface which of the two branches to render.
 */
data class AddWidgetButtonDisplay(
    val surface: AddWidgetButtonSurface,
) {
    /** True when the FAB should be drawn (the web not-editing branch). */
    val visible: Boolean get() = surface == AddWidgetButtonSurface.Visible
}

/**
 * Pure projection from the surface's `isEditing` input to its render-ready [AddWidgetButtonDisplay] — a 1:1
 * port of the web component's only decision: it returns null (renders nothing) in edit mode and the FAB
 * otherwise. Side-effect-free, so it is fully covered by the off-device unit gate.
 */
object AddWidgetButtonProjection {
    /**
     * Select the render-ready view for [isEditing]. In edit mode the surface is
     * [AddWidgetButtonSurface.Hidden] (web `return null`); otherwise it is [AddWidgetButtonSurface.Visible].
     */
    fun project(isEditing: Boolean): AddWidgetButtonDisplay =
        AddWidgetButtonDisplay(
            surface = if (isEditing) AddWidgetButtonSurface.Hidden else AddWidgetButtonSurface.Visible,
        )
}

/** Stable identifiers for the surface (P1/S11 diagnostics + the web-parity UI test tag). */
object AddWidgetButtonRegistration {
    /** Stable surface id. */
    const val ID: String = "add-widget-button"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "AddWidgetButton"

    /** Web-parity test tag — the web `data-testid="dashboard-add-widget-fab"`. */
    const val TEST_TAG: String = "dashboard-add-widget-fab"
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface slug — never the click
 * handler or any dashboard state — so a diagnostics line can never leak what the operator was viewing.
 */
object AddWidgetButtonDiagnostics {
    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-visible effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to AddWidgetButtonRegistration.SLUG))
    }
}
